DBZ-1218 Close JDBC connection in SnapshotReader for MySQL Connector

jgao54 opened the pull request:
The JDBC connection is established but never closed at the end of a snapshot execution resulting in resource leak.

wrp commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -719,6 +719,12 @@ protected void execute() {
+            try {
+                mysql.close();
+            } catch (SQLException e) {
+                // ignore
```
Rather than completely ignoring the failed close, it would be nice to at least log a message.

jgao54 replied:
added!

criccomini commented:
For reference, when Kafka connect rebalances during a snapshot, we see a deadlock where a connection appears to hold the lock in MySQL indefinitely. @jgao54 suspects that this patch might fix it

gunnarmorling commented:
Applied.

Thanks for the fix, @jgao54 and thanks for reviewing, @wrp!

> suspects that this patch might fix it

Interesting; I'd have expected that shutting down the connector implicitly closes the connection, but doing it explicitly is definitely the better way. Curious whether you'll still see this issues with the change applied or not.

jgao54 commented:
@gunnarmorling 
the issue seemed a bit more complicated --  not closing the jdbc connection wasn't the root cause.

Below is the set of steps that triggered this issue:
(1) to add a new table to dbz, we added a new connector via the REST api with config for this new table
(2) node A begins to snapshot this new table
(3) while node A is snapshotting, we changed the membership of the KC group by adding a new node, which triggered rebalancing of the nodes, and result in the new table to be assigned to node B instead.
(4) node B is now trying to snapshot this new table, but it can't make progress, and stuck on waiting for commit lock.

So before this patch,  we were seeing the following in the mysql processlist, and I thought that the connection was not closed:

```| 10 | dbz     | node_b_host:port | NULL | Query            |      90 | Waiting for commit lock                     | FLUSH TABLES WITH READ LOCK |         0 |             0 |```

After applying the change in this PR, we saw a slightly different message, but it was still stuck:

```| 10 | dbz     | node_b_host:port   | NULL  | Query            |     118 | Waiting for table flush                                       | FLUSH TABLES WITH READ LOCK             |         0 |             0 |```

It appears that the original `SELECT *` query from node A is still running long after the connector task is shut down. 

```| 6456721 | debezium | node_a_host:port | db | Query   | 5505 | Sending data | SELECT * FROM ***.*** | 12345 |             0 |```

We suspect (hopefully) it has something to do with our long mysql timeout setting... and are in the process of verifying this hypothesis. 

(created DBZ-1240 to report this issue)

gunnarmorling commented:
Thanks for reporting back, @jgao54!

> It appears that the original SELECT * query from node A is still running long after the connector task is shut down.
> We suspect (hopefully) it has something to do with our long mysql timeout setting.

jgao54 commented:
@gunnarmorling  We tuned the timeout config and it didn't solve the problem.... It turns out it's not a config issue... it's a problem with the fact that streaming resultset will wait for all records, even if the statement is closed (see [this bug report](https://bugs.mysql.com/bug.php?id=42929): `If the mysql protocol worked that way, the JDBC driver would use it. There's no way currently to close a result set "midstream", all of the rows have to be read off the wire, so it's going to be O(n). `).  Basically, if there is a rebalance happening during a snapshot, it's going to hang until the resultset has been fully streamed.

jpechane commented:
@jgao54 Have you tried to reduce `min.row.count.to.stream.results` config option?

gunnarmorling commented:
> streaming resultset will wait for all records, even if the statement is closed

Doh, that seems troublesome. Probably worth to re-open a discussion with the Connector/J driver maintainers?

One possible help could be the connector option we have for delaying snapshots for some configurable time; this has been added for similar issues where rebalancing during snapshotting would happen due to parallel deployment of other connectors.

jgao54 commented:
@jpechane  we use the default value of 1000. This is an issue with the jdbc lib unfortunately because once streaming starts it needs to stream all the results :( 

@gunnarmorling if we bounce the node which is doing the snapshotting, the connection will get terminated, that's basically how we've been "fixing" it right now. Not ideal. It seems it's something that should be fixed on the jdbc side, especially [the mysql doc](https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-reference-implementation-notes.html) actually suggests close() would work.

jpechane commented:
@jgao54 I must admit I am probably missing something from the picture. This setting should ensure that streaming is done in batches. So what exactly happens? Does it mean that even if result set is in forward only streaming mode the `connection.close()` blocks? Or it returns but the connection is not really closed?

jgao54 commented:
@jpechane  if my understanding is correct, `min.row.count.to.stream.results` whether or not the table(s) should be streamed, not how big the batch size is for streaming.

In fact, the jdbc driver doesn't have a "stream-in-batch" feature afaik. It will either (1) stream them one-by-one or (2) store the resultset in memory or (3) store the entire result in a temp table on the server side, and then batch fetch them based on a specified size (`stmt.setFetchSize(100)`).

jgao54 commented:
> Does it mean that even if result set is in forward only streaming mode the connection.close() blocks? Or it returns but the connection is not really closed?

It's blocking on `statement.close()` or `resultset.close()` afaik, likely the latter.

gunnarmorling commented:
> It seems it's something that should be fixed on the jdbc side, especially the mysql doc actually suggests close() would work.

Ok, could you perhaps open an issue in the MySQL JDBC connector then?
